pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +181,26 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, repo_id, file_obj):
+        upload_fts = []
+
+        if isinstance(file_obj, str):
+            file_obj = open(file_obj, "rb")
+
+        checksum = hashlib.sha256()
+        size, offset = 0, 0
+        data = file_obj.read(self._CHUNK_SIZE)
+        LOG.info("Uploading %s to repo %s", file_obj, repo_id)
+        while data:
+            checksum.update(data)
+            size += len(data)
+            upload_fts.append(self._do_upload(data, upload_id, offset))
+            offset += self._CHUNK_SIZE
+            data = file_obj.read(self._CHUNK_SIZE)
+        file_obj.close()
```

Comment:
Hmm... not certain this is correct.  file_obj could be coming from the caller.  Are you sure they wanted their file closed?  Couldn't it have some undesirable side-effect for them...

Yeah, I'm pretty sure this is a bad idea if you simply consider the use-case: you have an open file and you want to upload the same thing to N repositories.  With this close here, the caller would have to open the same file N times for some reason.

I think you should avoid closing the file_obj when it was passed by the user.

Reply:
hm, makes sense. 
Then I guess the right way is to make the upload_api only accepts file object, either it's an io.StringIO object or an opened file, then users could have more freedom.

Reply:
Okay, Then I think if users pass a string to the upload, then it should be closed.